src/vm/eventpipesession.cpp

```diff
@@ -81,12 +248,167 @@ EventPipeSessionProvider* EventPipeSession::GetSessionProvider(EventPipeProvider
+    EventPipeConfiguration &configuration,
+    LARGE_INTEGER stopTimeStamp)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+        PRECONDITION(EventPipe::IsLockOwnedByCurrentThread());
+    }
+    CONTRACTL_END;
+
+    if (m_pFile == nullptr)
+        return true;
+
+    // CrstHolder _crst(&m_lock);
+    m_pBufferManager->WriteAllBuffersToFile(m_pFile, configuration, stopTimeStamp);
+    return !m_pFile->HasErrors();
+}
+
+bool EventPipeSession::WriteEvent(
+        Thread *pThread,
+        EventPipeEvent &event,
+        EventPipeEventPayload &payload,
+        LPCGUID pActivityId,
+        LPCGUID pRelatedActivityId,
+        Thread *pEventThread,
+        StackContents *pStack)
+{
+    // CrstHolder _crst(&m_lock);
+    // TODO: Filter events specific to "this" session.
+    return m_pBufferManager->WriteEvent(pThread, *this, event, payload, pActivityId, pRelatedActivityId);
+}
+
+void EventPipeSession::WriteEvent(
+    EventPipeEventInstance &instance,
+    EventPipeConfiguration &configuration)
+{
+    if (m_pFile == nullptr)
+        return;
+
+    // CrstHolder _crst(&m_lock);
+    m_pFile->WriteEvent(instance, configuration);
+}
+
+EventPipeEventInstance *EventPipeSession::GetNextEvent()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+        PRECONDITION(!EventPipe::IsLockOwnedByCurrentThread());
+    }
+    CONTRACTL_END;
+
+    // CrstHolder _crst(&m_lock);
+    return m_pBufferManager->GetNextEvent();
+}
+
+void EventPipeSession::Enable(EventPipeProviderCallbackDataQueue *pEventPipeProviderCallbackDataQueue)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+        // Lock must be held by EventPipe::Enable.
+        PRECONDITION(EventPipe::IsLockOwnedByCurrentThread());
+    }
+    CONTRACTL_END;
+
+    if (m_SessionType == EventPipeSessionType::IpcStream)
+        CreateIpcStreamingThread();
+}
+
+void EventPipeSession::Disable(
+    EventPipeConfiguration &configuration,
+    LARGE_INTEGER stopTimeStamp,
+    EventPipeProviderCallbackDataQueue *pEventPipeProviderCallbackDataQueue)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+        // Lock must be held by EventPipe::Disable.
+        PRECONDITION(EventPipe::IsLockOwnedByCurrentThread());
+    }
+    CONTRACTL_END;
+
+    if (m_pFile == nullptr)
+        return;
+
+    // CrstHolder _crst(&m_lock);
+
+    // Disable streaming thread
+    if ((m_SessionType == EventPipeSessionType::IpcStream) && m_ipcStreamingEnabled)
+    {
+        // Reset the event before shutdown.
+        m_threadShutdownEvent.Reset();
+
+        // The IPC streaming thread will watch this value and exit
+        // when profiling is disabled.
+        m_ipcStreamingEnabled = false;
+
+        // Wait for the sampling thread to clean itself up.
+        m_threadShutdownEvent.Wait(INFINITE, FALSE /* bAlertable */);
```

Comment:
This is where you could have a deadlock.
Suppose one thread reach here and is waiting for the streaming thread to complete, and the streaming thread cannot complete because it cannot acquire the EventPipe::GetLock(), both thread will not be able to proceed and you have a deadlock.

Reply:
Yup, that's what I'm seeing.